Kudu.Services/Deployment/PushDeploymentController.cs

```diff
@@ -138,6 +140,164 @@ public class PushDeploymentController : ApiController
             }
         }
 
+        [HttpPost]
+        [HttpPut]
+        public async Task<HttpResponseMessage> OnePushDeploy(
+            [FromUri] bool isAsync = false,
+            [FromUri] string author = null,
+            [FromUri] string authorEmail = null,
+            [FromUri] string deployer = OneDeploy,
+            [FromUri] string message = DefaultMessage)
+        {
+            using (_tracer.Step("OnePushDeploy"))
+            {
+                var response = Request.CreateResponse();
+                //Checks the "type" query parameter, defaults to "static" if no query parameter is provided 
+               var deployType = Request.RequestUri.ParseQueryString()["type"];
+                if (string.IsNullOrWhiteSpace(deployType))
+                {
+                    deployType = "static";
+                }
+                var website_framework = (System.Environment.GetEnvironmentVariable("website_framework".ToLower())).ToLower();
+                //Checks the "path" queary parameter, path root at /wwwroot TBD if need to change to /home
+                string path = Request.RequestUri.ParseQueryString()["path"];
+                //Checks the "clean" query parameter, defauts to False if no query parameter is provided
+                var cleanQuery = Request.RequestUri.ParseQueryString()["clean"];
+                bool isCleanDeployment = Convert.ToBoolean(cleanQuery);
+                //Checks the "async" query parameter, defaults to false if no query parameter is provided
+                var asyncQuery = Request.RequestUri.ParseQueryString()["async"];
+                isAsync = Convert.ToBoolean(asyncQuery);
+
+                var deploymentInfo = new ZipDeploymentInfo(_environment, _traceFactory)
+                {
+                    AllowDeploymentWhileScmDisabled = true,
+                    Deployer = deployer,
+                    WatchedFilePath = Path.Combine("WEB-INF", "web.xml"),
+                    IsContinuous = false,
+                    AllowDeferredDeployment = false,
+                    IsReusable = false,
+                    CleanupTargetDirectory = isCleanDeployment,
+                    TargetChangeset = DeploymentManager.CreateTemporaryChangeSet(message: "Deploying from pushed file"),
+                    CommitId = null,
+                    RepositoryType = RepositoryType.None,
+                    Fetch = LocalAppFetch,
+                    DoFullBuildByDefault = false,
+                    Author = author,
+                    AuthorEmail = authorEmail,
+                    Message = message
+                };
+
+                switch (deployType)
+                {
+                    case "war":
```

Comment:
>"war": [](start = 25, length = 6)

Can you make an exhaustive list of all possibly scenarios that can ever happen (sync/async, error/no error/ etc) when using type=war and share it here as a comment (for now). Then we will do a similar exercise for other types.

Reply:
async=false : Complete
aysnc=true : Preliminary testing shows that it works as intended. More testing will have to be done.
clean=true : Complete
clean=false : Doesn't seem to work at the moment, I'll look into where deploymentInfo.CleanupTargetDirectory is referenced and investigate from there
path : Code currently does not support path for type = war. This can be easily changed though.
Errors : HTTP 400 when trying to deploy type=war to a Java SE webapp. If path is allowed for type=war, a HTTP 400 will be returned if the customer tried to specify a filename that does not have the .war extension. We will not be able to ensure that the file being deployed by the customer is a war file. 


---
In reply to: [455295447](https://github.com/shrishrirang/kudu/pull/2#discussion_r455295447) [](ancestors = 455295447)

Reply:
Some test cases: 
- POST with missing body
- POST with empty body
- path=invalid Linux path
- invalid Content-type
- invalid zip file (in case of wardeploy)
- multiple onedeploy at a time (Invoke onedeploy. Make it take longer by putting an artificial sleep, either hardcode or based on presence of a file, and then invoke the second and see what happens. Whether it fails with appropriate error message or not when type=war? See if the behavior matches wardeploy.
- what else?

---
In reply to: [455354445](https://github.com/shrishrirang/kudu/pull/2#discussion_r455354445) [](ancestors = 455354445,455295447)

Reply:
When you get a chance to check these scenarios, let's capture the summary.

---
In reply to: [456011982](https://github.com/shrishrirang/kudu/pull/2#discussion_r456011982) [](ancestors = 456011982,455354445,455295447)